### PR TITLE
[virt_autotest] fix up set_vcpus failure for fv guest on sles15sp4 xen host

### PR DESCRIPTION
**Pull request by guoxuguang**
Refer to the latest OSD sles15sp4 build 29.1 results, figure out the new changes about set_vcpus failure for fv guest on sles15sp4 xen host. 
After deployed all sles12+ fv guest (e.g sles15sp4 fv guest) with 2 vcpu as the default, the try to setvcpus as 2 again to ensure deployed fv guest (e.g sles15sp4 fv guest) CPU count is 2:
$ virsh setvcpus --domain sles-15-sp4-64-fv-def-net --count 2 --live
Figure out the following error:
error: internal error: Failed to set vcpus for domain '26' with libxenlight
But, this step was successful with the sles15sp4 build26.1 - https://openqa.suse.de/tests/6888588#step/hotplugging/193

Please refer to [bsc#1188898](https://bugzilla.suse.com/show_bug.cgi?id=1188898)  for more details. 

Create this PR to workaround this set_vcpus failure for fv guest on sles15sp4 xen host

- Verification run: 
[gi-guest_developing-on-host_developing-xen](https://openqa.suse.de/tests/6986387#)
[gi-guest_sles12sp5-on-host_developing-xen](https://openqa.suse.de/tests/7030709#)
[gi-guest_sles15sp3-on-host_developing-xen](https://openqa.suse.de/tests/7030710#)
[gi-guest_developing-on-host_sles15sp3-xen](https://openqa.suse.de/tests/6990920#)

### CODE REVIEW

**Comment by guoxuguang** [comment]
@pdostal @grisu48 please help me review this PR. Also, help me make sure if this PR worked on your QAM environment.

**Comment by Julie-CAO** [comment]
bsc#1188898 has been fixed, why do we skip testing and make it soft fail?

**Comment by guoxuguang** [comment]
> bsc#1188898 has been fixed, why do we skip testing and make it soft fail?

Good question,  bsc#1188898 has been fixed from our developer now. But, this product issue still reproduce on the latest 15sp4 **build29.1** as my mentioned. 
Moreover, current this product issue still block the following osd test on sles15sp4 build29.1:
gi-guest_developing-on-host_developing-xen
gi-guest_sles12sp5-on-host_developing-xen
gi-guest_sles15sp3-on-host_developing-xen

Also, compare 15sp4 build26.1 and build29.1 failure result for this bsc#1188898, there was a little bit changes - fail at **Ensure guest CPU count is 2** step(build29.1) without **# Add 1 CPU** step (build26.1). 

So, we have to create this PR to deal with this a little change for bsc#1188898 on sles15sp4 build29.1 to skip testing and make it soft fail again.

**Review by pdostal — CHANGES_REQUESTED** [review]
I don't understand why are you moving so much of code, I suggest something like:
```perl
sub test_add_vcpu {
  ...
  if (is_xen_host && ($sles_running_version eq '15' && $sles_running_sp eq '4') && $gues =~ /sles12sp4PV/) {
    record_info 'Skip vCPU hotplugging', 'bsc#XXX vCPU hotplugging does no work on SLE 12-SP4 PV guest on SLE15-SP4 host';
    return;
  }
  ...
```

And of course, there is the question if this workaround is even needed as it's a bug in latest product which should be fixed soon.

**Comment by guoxuguang** [comment]
> I don't understand why are you moving so much of code, I suggest something like:
> 
> ```perl
> sub test_add_vcpu {
>   ...
>   if (is_xen_host && ($sles_running_version eq '15' && $sles_running_sp eq '4') && $gues =~ /sles12sp4PV/) {
>     record_info 'Skip vCPU hotplugging', 'bsc#XXX vCPU hotplugging does no work on SLE 12-SP4 PV guest on SLE15-SP4 host';
>     return;
>   }
>   ...
> ```
> 

Cool! Good suggestion here. I will follow this suggestion to update code. 

> And of course, there is the question if this workaround is even needed as it's a bug in latest product which should be fixed soon.
Good question,  there was a long stroy for this bsc#1188898 issue. our developer had updated this issue as fixed early. But, this same issue still block our hotpluging tests fo all fv guest from build 16.3 to 31.2 on sles15sp4 xen host. So, we have to use with the workaround mode for this bsc#1188898 issue to continue our VT tests on OSD.

**Review by pdostal — APPROVED** [review]
Thank you, @guoxuguang, good to go from my side :rocket:

File: tests/virtualization/universal/hotplugging.pm [hunk 1]
```diff
@@ -136,14 +136,16 @@ sub test_add_vcpu {
     }
     return if (is_xen_host && $guest =~ m/hvm/i);    # not supported on HVM guest
 
+    if ($sles_running_version eq '15' && $sles_running_sp eq '4' && is_xen_host && is_fv_guest($guest)) {
+        record_info 'Skip vCPU hotplugging', 'bsc#1188898 vCPU hotplugging does no work on SLE FV guest on SLE15-SP4 host';
```
Comment (waynechen55) [inline tests/virtualization/universal/hotplugging.pm:140]: Why not record_soft_failure ?

File: tests/virtualization/universal/hotplugging.pm [hunk 2]
```diff
@@ -136,14 +136,16 @@ sub test_add_vcpu {
...
     assert_script_run("ssh root\@$guest nproc | grep 2", 60);
     # Add 1 CPU
-    if ($sles_running_version eq '15' && $sles_running_sp eq '4' && is_xen_host && is_fv_guest($guest)) {
-        record_soft_failure('bsc#1188898 Failed to set live vcpu count on fv guest on 15-SP4 Xen host');
-    }
-    elsif ($sles_running_version eq '15' && $sles_running_sp eq '3' && is_xen_host && is_fv_guest($guest)) {
+    if ($sles_running_version eq '15' && $sles_running_sp eq '3' && is_xen_host && is_fv_guest($guest)) {
         record_soft_failure('bsc#1180350 Failed to set live vcpu count on fv guest on 15-SP3 Xen host');
     }
     else {
```
Comment (waynechen55) [inline tests/virtualization/universal/hotplugging.pm:151]: What about move
```
# Ensure guest CPU count is 2
die "Setting vcpus failed" unless (set_vcpus($guest, 2));
assert_script_run("ssh root\@$guest nproc | grep 2", 60);
```
into the "else" branch ? Then you do not need to change anything else.

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:151]: Refer to the test_add_vcpu {} roadmap, there should be **# Ensure guest CPU count is 2** firstly and then **# Add 1 CPU**.

Comment (waynechen55) [reply tests/virtualization/universal/hotplugging.pm:151]: For the test, Yes. But if the test does not need to run, then "Ensure guest CPU count is 2" is not needed either.

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:151]: yes, we need to figure out the easy and simple way to skip this `test_add_vcpu {}` on sles xen host with FV guest.  I totally agree on your POV here. But, there were more `if` and `else` branch under `test_add_vcpu {}`.

Comment (waynechen55) [reply tests/virtualization/universal/hotplugging.pm:151]: I see. But I think, at least, we can make the if...elsif...else better by making the change I suggested above. Maybe @pdostal can provide relevant input here about how this looks like from their testing POV. If he does not object either, we can make the change ?

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:151]: Awesome! yes, I totally agree on your suggestion as my mentioned. also, we'd better to figure out the easy and simple way to maintain this test module with QAM together. 
btw, regarding the bsc#1188898 issue on sles15sp4 was a little bit more special as my mentioned. Due to compare with the latest sles15sp4 build 31.2 and build 26.1,  `bsc#1188898` had a little changed for `bsc#1180350`. if still used with `if... elsif ...else` means that move so much of codes to keep the same  logical structure for this test module POV.

Comment (waynechen55) [reply tests/virtualization/universal/hotplugging.pm:151]: I  understand your concern. You want to keep the original test structure. Right ?

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:152]: due to the current bsc#1188898 issue on sles15sp4, plan to keep the original test structure for this hotplugging test module. 
For now, I am not clear if there was still new changes for **bsc#1188898** on sles15sp4 again during `hotplugging::test_add_vcpu {}`. 
Anyway, hope our developer fix this **bsc#1188898** on sles15sp4 quickly.

Comment (waynechen55) [reply tests/virtualization/universal/hotplugging.pm:152]: I do not think this involves any specific bug or bug fix. Basically the "else" branch should include everything needs to be set and test. If any test scenario should be skipped then it should be in another "if" or "elsif" branch to skip it. Currently the "Ensure guest CPU count is 2" part wanders around instead of  being the pre-conditions to sit together with the actual test part. In other words, there is no point to "ensure anything" if "actual test part should be skipped".

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:152]: yes, I see.  As the purpose for this PR to skip `test_add_vcpu {}` for **bsc#1188898** on sles15sp4. 
Also, Due to currently the "Ensure guest CPU count is 2" part wanders around instead of being the pre-conditions to sit together with the actual test part. maybe we can continue this season on ticket [Separate hotplugging.md into _nic, _cpu, _mem files](https://progress.opensuse.org/issues/96230).

File: tests/virtualization/universal/hotplugging.pm (hunk 1 above)
Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm]: yup, update to use with record_soft_failure 
[
The latest validation run](https://openqa.suse.de/tests/7030710#)

File: tests/virtualization/universal/hotplugging.pm [hunk 3]
```diff
@@ -136,14 +136,16 @@ sub test_add_vcpu {
...
+        record_soft_failure('bsc#1188898 Failed to set live vcpu count on fv guest on 15-SP4 Xen host');
+        return;
+    }
+
     # Ensure guest CPU count is 2
     die "Setting vcpus failed" unless (set_vcpus($guest, 2));
     assert_script_run("ssh root\@$guest nproc | grep 2", 60);
     # Add 1 CPU
-    if ($sles_running_version eq '15' && $sles_running_sp eq '4' && is_xen_host && is_fv_guest($guest)) {
-        record_soft_failure('bsc#1188898 Failed to set live vcpu count on fv guest on 15-SP4 Xen host');
```
Comment (pdostal) [inline tests/virtualization/universal/hotplugging.pm:144]: Why did you remove `bsc#1188898`?

Comment (guoxuguang) [reply tests/virtualization/universal/hotplugging.pm:144]: move bsc#118889 at line 139 to 143 now
